File: frontend/ui/src/utils/compressImage.js
import imageCompression from "browser-image-compression";

/**
 * Compresses an image by resizing it and reducing its quality.
 *
 * @param {File} file - The image file to compress.
 * @param {number} maxWidth - Maximum allowed width for the image.
 * @param {number} maxHeight - Maximum allowed height for the image.
 * @param {number} compressionQuality - Compression quality (0 to 1, higher means better quality).
 * @returns {Promise<File>} - The compressed image file.
 */
async function compressImage(
	file,
	maxWidth = 899,
	maxHeight = 1599,
	compressionQuality = 0.85
) {
	try {
		// Create an image object to get the original dimensions
		const originalImage = await new Promise((resolve, reject) => {
			const img = new Image();
			img.onload = () => resolve(img);
			img.onerror = reject;
			img.src = URL.createObjectURL(file);
		});

		const actualWidth = originalImage.width;
		const actualHeight = originalImage.height;
		console.log("Original Dimensions:", `${actualWidth}x${actualHeight}`);

		const imgRatio = actualWidth / actualHeight;
		const maxRatio = maxWidth / maxHeight;

		let newWidth = actualWidth;
		let newHeight = actualHeight;

		// Adjust dimensions to fit within the max dimensions while maintaining aspect ratio
		if (actualWidth > maxWidth || actualHeight > maxHeight) {
			if (imgRatio < maxRatio) {
				// Adjust width according to maxHeight
				const scaleFactor = maxHeight / actualHeight;
				newWidth = Math.floor(scaleFactor * actualWidth);
				newHeight = maxHeight;
			} else {
				// Adjust height according to maxWidth
				const scaleFactor = maxWidth / actualWidth;
				newHeight = Math.floor(scaleFactor * actualHeight);
				newWidth = maxWidth;
			}
		}

		console.log("Resized Dimensions:", `${newWidth}x${newHeight}`);

		// Prepare options for image compression
		const options = {
			maxSizeMB: 1, // Set max file size in MB
			maxWidthOrHeight: Math.max(newWidth, newHeight), // Set max width or height
			useWebWorker: true, // Use Web Workers for faster compression
			initialQuality: compressionQuality, // Quality of the output image (0 to 1)
		};

		console.log(
			"Original File Size:",
			(file.size / 1024 / 1024).toFixed(2),
			"MB"
		);

		// Compress the image
		const compressedFile = await imageCompression(file, options);

		console.log(
			"Compressed File Size:",
			(compressedFile.size / 1024 / 1024).toFixed(2),
			"MB"
		);

		// Return the compressed image
		return compressedFile;
	} catch (error) {
		console.error("Image compression failed:", error);
		throw error;
	}
}

export default compressImage;

File: backend/attendify/utils/compress_image.py
from PIL import Image
import os

def compress_image(input_image_path, max_width=899, max_height=1599, compression_quality=85):
    """
    Compresses an image by resizing it and reducing its quality.

    Parameters:
        input_image_path (str): Path to the input image.
        max_width (int): Maximum allowed width for the image.
        max_height (int): Maximum allowed height for the image.
        compression_quality (int): Compression quality (0 to 100, higher means better quality).

    Returns:
        str: Path to the compressed image.
    """
    # Open the image using PIL
    image = Image.open(input_image_path)

    # Get original dimensions
    actual_width, actual_height = image.size
    print("Actual width:", actual_width)
    print("Actual height:", actual_height)
    img_ratio = actual_width / actual_height
    max_ratio = max_width / max_height

    # Adjust dimensions to fit within the max dimensions while maintaining aspect ratio
    if actual_height > max_height or actual_width > max_width:
        if img_ratio < max_ratio:
            # Adjust width according to maxHeight
            scale_factor = max_height / actual_height
            actual_width = int(scale_factor * actual_width)
            actual_height = max_height
        else:
            # Adjust height according to maxWidth
            scale_factor = max_width / actual_width
            actual_height = int(scale_factor * actual_height)
            actual_width = max_width

    # Resize the image
    resized_image = image.resize((actual_width, actual_height), Image.LANCZOS)

    # Generate the output image path
    base, ext = os.path.splitext(input_image_path)
    output_image_path = f"{base}_compressed{ext}"

    # Save the image with reduced quality (compression)
    resized_image.save(output_image_path, "JPEG", quality=compression_quality)

    return output_image_path


# Example Usage
if __name__ == "__main__":
    # Replace 'your_image.jpg' with the path to your image file
    # input_image = r"D:\Web Dev\Collaborate\Attendify\backend\attendify\utils\IMG20241119224436.jpg"
    input_image = r"D:\Web Dev\Collaborate\Attendify\backend\attendify\utils\rahul 28 mb.jpg"
    compressed_image_path = compress_image(input_image)
    print(f"Compressed image saved to: {compressed_image_path}")